updating domtrees dynamically, removing all unreachable blocks

yhls opened the pull request:
This all started with #29107, which (is at least one thing that) stands in the way of turning DCE back on globally (it was disabled in #29265). The problem is that when there is only one predecessor, replacing a phi node with its only possible value can result in a forward reference when this phi node is in an unreachable loop. This can cause problems with passes that process all IR, whether or not the code is dead.

We can remove dead loops by removing all unreachable blocks, instead of just removing blocks with no predecessors, which is what we currently do. To determine reachability, all that would be required is a depth-first traversal of the CFG, but as suggested by #29140, I have implemented an algorithm for updating the dominator tree of a CFG dynamically. I hear that this would be useful to have anyway. Just to be safe, I’ve also prevented a phi node from being replaced with its value if that value is defined after the phi node is.

The dynamic domtree algorithm basically works by figuring out if we can get away with doing nothing or recomputing only some of the immediate dominators when we add or delete an edge in the CFG, or if we have to recompute the whole domtree from scratch.

As this PR stands, domtrees are not updated dynamically yet, even though the dynamic algorithm has been implemented. First, I tried just computing the domtree from scratch when we decide which blocks are dead, which happens when an `IncrementalCompact` is created. I tried to get an idea of the performance impact by comparing how long Base and Stdlibs took to compile, but the compile times were pretty noisy and I couldn’t confidently notice any difference. Next, I tried adding a `DomTree` to every `CFG` and computing it every time a `CFG` is constructed. This appears to lengthen compile times by about 5% compared with master, and 3% compared with master with DCE on. The only times a new `CFG` is made from an old one are when we finish iterating over an `IncrementalCompact`, and in `domsort_ssa!`. Presumably we would be able to recover some performance by updating the domtree dynamically in these places.

I’ve also addressed a bug wherein references to dead blocks that have been removed remain in phi nodes.

yhls committed:
Decline to remove phi nodes when this might cause forward references

Phi nodes are optimized away when there is only one predecessor, but this can
cause problems in dead loops because forward references can be created, leading
to issues with optimization passes that look at all code, dead or not. This
fixes issue #29107 when DCE is turned on.

yhls committed:
Implement algorithm for updating dominator tree with CFG edge insertions and deletions

The DFS tree associated with a CFG now keeps track of postorder as well as
preorder numbers. The DFS tree, as well as the state associated with the SNCA
algorithm for finding (immediate) dominators is now stored in DomTree and
reused for Dynamic SNCA.

yhls committed:
Make domtree construction take `Vector{BasicBlock}` as input instead of CFG

This is in anticipation of domtrees being added to CFGs.

yhls committed:
Remove dead blocks as determined by reachability instead of number of predecessors

For now, just construct the domtree when we make an `IncrementalCompact` rather
than try to update it (the domtree) incrementally.

yhls committed:
Remove edges from dead blocks from phi nodes

yhls committed:
Add a domtree to every CFG, avoiding constructing them explicitly when needed

Every time a CFG is created, its corresponding dominator tree is as well.

yhls committed:
Enable CFG transforms; turn DCE back on globally

yhls commented:
Hmm, it seems to take forever on test/compiler/irpasses.jl, will investigate.

yhls commented:
Fixed. Turns out that was because I was verifying the domtree against output from the naive immediate dominator algorithm, which takes way too long on a big test case in test/compiler/irpasses.jl.

Keno commented on base/compiler/ssair/ir.jl:
```diff
@@ -52,8 +52,34 @@ struct CFG
     blocks::Vector{BasicBlock}
     index::Vector{Int} # map from instruction => basic-block number
                        # TODO: make this O(1) instead of O(log(n_blocks))?
+    domtree
```
More concrete type on this field?

yhls replied:
Right, I forgot to mention--domtree.jl is `include`d after ir.jl, but uses `CFG` and `BasicBlock`, which are defined in ir.jl, so we'll have to do some reorganizing for this to work.

Keno commented:
Amazing work (seriously, I'm quite impressed). However, before getting into the implementation, I think we have a more high-level problem: I don't think eliminating dead blocks in the `IncrementalCompact` constructor is enough. The issue that caused the original problem is that removing edges can turn instructions in unreachable blocks invalid and if the pass then goes to process them we can get into a bad state. Of course the dynamic domtree path is designed to address that, so we're almost there with this PR, but I don't think we can turn it on by default until we actually update it in IncrementalCompact whenever we remove an edge and kill all invalid blocks right then (or at least remove all contained instructions). Am I missing something?

yhls commented:
Removing dead blocks in the constructor is actually enough to fix #29107, because the particular thing that was going wrong (phi nodes being replaced if they only had one possible value, which can be bad in dead loops) was only ever done while iterating through an `IncrementalCompact`, while it looks like edges are currently only ever removed outside of them, during inlining. But I agree that there could potentially be a case where something goes wrong if we remove an edge while iterating through an `IncrementalCompact` (although the `kill_edge!` for `IncrementalCompact` objects is currently unused).

According to #29107, invalid IR in dead code is explicitly allowed (I think this is also true for LLVM). Do we want to keep this convention? If so, once we really do update the domtree dynamically, passes will be able to check if a block is dead or not instantaneously. Or do we actually want never to have any dead code at all?

yhls commented:
Wait, the `kill_edge!` for `IncrementalCompact` objects is used...

yhls committed:
Update domtrees dynamically when CFGs get modified, and fix bug in dynamic domtree implementation

yhls committed:
Fix possible merge cycle bug in `cfg_simplify!`

yhls committed:
Update domtree dynamically in `cfg_simplify`

yhls committed:
Fix infinite loop bug in `kill_edge!`

If a statement was `nothing`, `kill_edge!` would never move on from trying to
kill it because the index wasn't incremented.

yhls committed:
Fix bug in `kill_edge!` which caused edges to be removed from the wrong phi nodes

yhls committed:
Kill statements in blocks that are unreachable, add tests for killing edges

This change only affects statements that we have yet to encounter after killing
an edge, while iterating through `IncrementalCompact`. Statements in dead
blocks that come before the point at which the edge is killed are killed in
`kill_edge!`, when the edge is killed.

yhls committed:
Fix bug in which uses of SSA values processed in `just_fixup!` were not counted

This fixes #29253, which was caused by `simple_dce!` erroneously erasing SSA
values that did not appear to be used, because these uses were only discovered
in `just_fixup!` at the end of iterating over an `IncrementalCompact`.

yhls committed:
Move `BasicBlock` to separate source file

This is so we can add type declarations to fields in ir.jl that are domtrees,
by breaking the dependency loop between domtree.jl (uses basic blocks but
defines domtrees) and ir.jl (uses domtrees but defined basic blocks).

yhls commented:
Ok, domtrees are updated dynamically now, and I fixed the bugs that cropped up when I turned DCE back on. Here's a summary of changes in this PR:
- DCE is turned back on (reverting #29265), and all unreachable blocks are removed (not just ones with no predecessors), as determined when `IncrementalCompact`s are constructed
- if `kill_edge!` makes blocks unreachable, all their statements are now killed immediately (set to `nothing` or `ReturnNode()`--this should prevent passes from seeing invalid statements in dead blocks
- phi node values no longer get forwarded to their uses in the case that there's only one incoming edge, if this value is coming from after the phi node (this fixes #29107, although is just to be safe because removing all unreachable blocks also fixes it)
- unreachable blocks are no longer referenced in phi nodes
- fixed #29253, an infinite loop in `kill_edge!`, block merging cycles in `cfg_simplify!`, and missing `used_ssas` counts

Some changes to the code:
- every `CFG` now contains a `DomTree`, which must be kept up to date
- `DomTree`s are no longer explicitly passed to functions like `domsort_ssa!`
- `construct_domtree` now takes `Vector{BasicBlock}` instead of `CFG`
- `DFSTree`s now keep track of postorder numbers as well as preorder numbers, and are stored in every `DomTree`

Anything that modifies the CFG will have to keep the domtree up to date as well. `kill_edge!(IncrementalCompact, ...)` does this already, but otherwise, `cfg_insert_edge!`, `cfg_delete_edge!`, `domtree_insert_edge!`, `domtree_delete_edge!`, `rename_blocks!`,  and `combine_blocks!` can be used. If many edge insertions or deletions are being done, it may be faster (and simpler) just to reconstruct the domtree from scratch, because the algorithm resorts to reconstructing the whole domtree if the change affects the preorder number of the blocks. (This is why the domtree is recomputed from scratch in `domsort_ssa!` and `batch_inline`.)

yhls commented:
Here's a rough measurement of impact on performance: in one run, master took a total of 67.2 seconds to compile the sysimage, and with this PR, 70.1 seconds (104%). Note that master with the modification that `allow_cfg_transforms=true` took 68.7 seconds (102%).

StefanKarpinski commented:
Perhaps it’s obvious or stated somewhere I’m missing, but what is the benefit of this change?

ViralBShah commented:
As @vchuravy explained on slack: Cleaner IR, better readability, and better for tools like XLA.jl that operate on typed IR

yhls commented:
Just edited some commits to avoid changing some structs into mutable structs.

vchuravy commented on base/compiler/ssair/domtree.jl:
```diff
@@ -1,63 +1,671 @@
+length(D::DFSTree) = length(D.from_pre)
+
+function DFS!(D::DFSTree, blocks::Vector{BasicBlock})
+    copy!(D, DFSTree(length(blocks)))
```
Is this `empty!`/`resize!`?

yhls replied:
It would be a `resize!` if l

vchuravy reviewed (approved):
LGTM from my side, although I would like @keno to sign off as well, since he has a deeper understanding
of that part of the code.

yhls committed:
Fix small bug in #36684

PR #36684 changes `iterate(IncrementalCompact)` to return an extra index, but
leaves its arguments unchanged. However, the PR decremented the index argument
in a particular recursive call to `iterate`. This caused `iterate` not to
recognise that it was done when `allow_cfg_transforms` was turned on.

yhls commented:
Rebased onto master, got things working with DCE turned on again with a small bugfix for #36684

yhls commented:
I've separated the freestanding bugfixes into #36888

vchuravy commented:
Replaced by #37882
